metal/resource_metal_port.go

```diff
@@ -0,0 +1,342 @@
+		}
+	}
+
+	return port, nil
+}
+
+func getSpecifiedNative(d *schema.ResourceData) string {
+	nativeRaw, nativeOk := d.GetOk("native_vlan_id")
+	specifiedNative := ""
+	if nativeOk {
+		specifiedNative = nativeRaw.(string)
+	}
+	return specifiedNative
+}
+
+func getCurrentNative(p *packngo.Port) string {
+	currentNative := ""
+	if p.NativeVirtualNetwork != nil {
+		currentNative = p.NativeVirtualNetwork.ID
+	}
+	return currentNative
+}
+
+func attachedVlanIds(p *packngo.Port) []string {
+	attached := []string{}
+	for _, v := range p.AttachedVirtualNetworks {
+		attached = append(attached, v.ID)
+	}
+	return attached
+}
+
+func specifiedVlanIds(d *schema.ResourceData) []string {
+	vlanIdsRaw, vlanIdsOk := d.GetOk("vlan_ids")
+	specified := []string{}
+	if vlanIdsOk {
+		specified = convertStringArr(vlanIdsRaw.([]interface{}))
+	}
+	return specified
+}
+
+type portVlanAction func(*packngo.PortAssignRequest) (*packngo.Port, *packngo.Response, error)
+
+func processVlansOnPort(portId string, vlanIds []string, f portVlanAction) error {
+	par := packngo.PortAssignRequest{PortID: portId}
+	for _, vId := range vlanIds {
+		par.VirtualNetworkID = vId
+		_, _, err := f(&par)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+/*
+Create and Update will be probably identical in case of metal_port
+
+I figured we should process port action in order:
+ - native VLAN detachment
+ - VLANs detachment
+ - bond break
+ - l2/l3 conversion
+ - bond creation
+ - VLANs assignment
+ - native VLAN assignment
+*/
+
+func resourceMetalPortCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*packngo.Client)
+
+	port, err := getPortByResourceData(d, client)
+	if err != nil {
+		return err
+	}
+
+	// Constraint: Only bond ports have layer2 mode
+	l2, l2Ok := d.GetOk("layer2")
+	if port.Type != "NetworkBondPort" && l2Ok {
+		return fmt.Errorf("layer2 flag can be specified only for bond ports")
+	}
+
+	// remove native vlan
+	currentNative := getCurrentNative(port)
+	specifiedNative := getSpecifiedNative(d)
+	if (currentNative != specifiedNative) && currentNative != "" {
+		_, _, err = client.DevicePorts.UnassignNative(port.ID)
+		if err != nil {
+			return err
+		}
+	}
+
+	// remove vlans
+	vlansToRemove := difference(
+		attachedVlanIds(port),
+		specifiedVlanIds(d),
+	)
+	err = processVlansOnPort(port.ID, vlansToRemove, client.DevicePorts.Unassign)
+	if err != nil {
+		return err
+	}
+
+	// disbond
+	bonded, bondedOk := d.GetOk("bonded")
+	if bondedOk && !bonded.(bool) && port.Data.Bonded {
+		port, _, err = client.DevicePorts.Disbond(port, false)
+		if err != nil {
+			return err
+		}
+	}
+
+	// convert to layer2
+	isLayer2 := contains(l2Types, port.NetworkType)
+
+	if l2Ok && l2.(bool) && !isLayer2 {
+		port, _, err = client.Ports.ConvertToLayerTwo(port.ID)
+		if err != nil {
+			return err
+		}
+	}
+
+	// convert to layer3
+	if l2Ok && !l2.(bool) && isLayer2 {
+		ips := []packngo.AddressRequest{
+			{AddressFamily: 4, Public: true},
```

Comment:
In the future it would be nice to provide the AddressRequest arguments for selecting existing addresses and creating EIPs through this `metal_port` resource.

Potentially avoiding the need for `metal_ip_address_attachment` resources.  That can be figured out in the future. We would have to consider what to do when the two resources are in conflict (metal_ip_address_attachment creates attachments not found in the metal_port list. Does the metal_port resource become the defacto way to specify all of these features?).

Reply:
I see that the L3 conversion has the possibility to specify address requests, but I'd leave it to a PR in the future, separately form the initial introduction of metal_port. I feel there will be other issues to focus on. 

Also, it's not possible to specify IPs in the console when converting to L3.